### ISSUE TITLE
Resolve node_modules path in compiler

### ISSUE THREAD

**Issue by leTrice** [juanfranblanco/vscode-solidity#31]
The compiler can't resolve the node_modules path, e.g. as follows:
```
import "node_modules/zeppelin-solidity/contracts/math/SafeMath.sol";
````

I have to use the complete relative path to modules, and the vscode-solidity compiler will work:
```
import "../../node_modules/zeppelin-solidity/contracts/math/SafeMath.sol";
````

Did i miss something in configuration, or will this be fixed in an future version?

Best regards
Patrice

**Comment by juanfranblanco** [comment]
Hi,

import "node_modules.." will not resolve anything, if there is not a . prefix. A dependency module it is assumed (as per lib).

A project file for this could be setup, in which it the library folders is:
"node_modules/zeppelin-solidity/"

``` import "../../node_modules/zeppelin-solidity/contracts/math/SafeMath.sol"; ```

This should work, I'll have a look at both.

**Comment by juanfranblanco** [comment]
![image](https://user-images.githubusercontent.com/562371/30062712-03a5f84e-9244-11e7-9888-cababd66330c.png)

Example of resolving a relative path and autocomplete

**Comment by juanfranblanco** [comment]
![image](https://user-images.githubusercontent.com/562371/30062845-855703ba-9244-11e7-9f0c-feff1fb58843.png)

Example of a directory

**Comment by juanfranblanco** [comment]
![image](https://user-images.githubusercontent.com/562371/30062969-ecf39682-9244-11e7-8d9c-88f80588666a.png)

If you rename "node_modules" to "lib" and "zeppelin_solity/contracts" to "zeppelin_solidity/src" then you can achieve the above

**Comment by juanfranblanco** [comment]
The TODO is to add a set of folders in the project file for both the contracts and libraries, so they can be resolved, and don't force a structure.

**Comment by SCBuergel** [comment]
@juanfranblanco both `import "../../node_modules.....` as well as copying or sym-linking seems more like a hack. I see many projects relying on truffle. Supporting these projects in an IDE-agnostic fashion would really help here. Therefore I'd suggest to be compatible with truffle and thus resolve everything in the `node_modules` folder by default.
The bigger picture is that I'm trying to limit the copy-pasting of Zeppelin libraries - and encourage others to follow good practices as well. That requires tools to be sufficiently flexible hence including the `node_modules` folder would bring us quite a bit forward towards more secure smart contracts in general.

**Comment by juanfranblanco** [comment]
@SCBuergel  the project structure at the moment is based on an ERP which matches the structure of Dappsys, which they have the modules in the "lib" folder.

**Comment by awgneo** [comment]
@juanfranblanco Truffle is a project with over 2,300 stars compared to dappsys' 127. Package management is still a moving target in the Ethereum world but seems better defined by Truffle. Even the ethpm spec doesn't seem to be as popular as npm for the time being.

**Comment by juanfranblanco** [comment]
@awgneo 2 completely different things (BTW I work in Consensys with Tim :) ). Dappsys is a library of solidity smart contracts like Zeppelin. Truffle allows you to deploy, compile, migrate, etc, etc :).

And yes compatibility with everything :)

**Comment by awgneo** [comment]
Great to hear. I know these are different things. However, conceptually they are similar. Truffle has the ability to install (truffle install) packages from ethpm, for which it searches an installed_contracts directory. Truffle also searches node_modules for contract imports. Finally, it searches its own contracts directory for local contracts, libraries, etc. The lib directory combined with GitHub submodules is essentially just another way to import other people's "packages" or libraries. So many things to support indeed.

**Comment by ezynda3** [comment]
How about adding a setting for adding custom library paths so folks can just add whatever they like?

**Comment by juanfranblanco** [comment]
Thanks @ezynda3  Yes that was mainly the direction, your project file should indicate where the dependencies reside, you can cascade later on different configurations. The issue is that now it is "hardcoded" to lib. If you have it a single path as a user setting, it becomes global and forces you in a specific library direction, but yes having an array of different paths would work for all. Great point!

**Comment by joejordan** [comment]
I came across this issue when trying to set up zeppelin-solidity "the right way". Currently, the right way gives you "file not found" errors in the IDE on the imports, even though Truffle compiles the project fine. 

If you copy the zeppelin-solidity/contracts folder to the root project directory it fixes the red squiggles, but better would be to support node_modules by default.

**Comment by qbit86** [comment]
> my TODO as per the other comment is to be able to create a project file, and add a user variable which will allow "node_modules" (or anything you fancy) being a "library" folder.

Any news on this? Is there any way to add `../../node_modules/` to probing paths?

**Comment by juanfranblanco** [comment]
I will dedicate next week or after to the solidity extension, this will be one of the priorities

**Comment by jtakalai** [comment]
Short: DappSys is a cool project, but OpenZeppelin recommends truffle.

Long: DappSys is a cool project, BUT:
* they recommend using [Nix](https://nixos.org/) package manager. Why pull in an exotic package manager when there already is npm? Because they're not Javascript? Nix already exists on every machine running NixOS - The Purely Functional Linux Distribution. I know mine doesn't.
* it's yet another (incompatible) Solidity framework. If you're familiar with JS, you choose truffle (or embark). If you're a Pythonista, you maybe choose populus. DappSys is neither, so it doesn't have an existing robust and well-known (shell) scripting platform for deployment etc. automation.
* web3.js is something of a reference implementation of web3, and much of the ecosystem seems to be leaning on JS. I don't love JS, but that's my reason of choosing it over, say, python in Solidity dev. Solidity VSCode plugin would do well to follow that suit and support the truffle way, in addition to possibly cooler things out there.

**Comment by ImmuneGit** [comment]
I see that the problem is still presents. VSCode can't find a path to zeppelin's contracts when I write like this: import "zeppelin-solidity/contracts/ownership/Ownable.sol".
<img width="1152" alt="screen shot 2018-03-18 at 12 26 45 pm" src="https://user-images.githubusercontent.com/25203571/37564469-aec10416-2aa7-11e8-8948-8ae27b565ea5.png">

**Comment by juanfranblanco** [comment]
Yes, next on the list. There is a simple pull to fix this.. but not complete, as it breaks previous stuff.

**Comment by juanfranblanco** [comment]
Actually, it did not break much the pull, many thanks to @bramhoven for being a total star and starting the work on this. The settings are now configurable so everyone can work in their preferred manner. IMHO having a standard project file, that list dependencies, versions etc it is the best way forward as envision a couple of years ago.

**Comment by qbit86** [comment]
@juanfranblanco 

>  The settings are now configurable

How can I configure VS Code extension 0.0.35 so that it compiles this line as `truffle` does?
```
import "zeppelin-solidity/contracts/ownership/Ownable.sol";
```

Now it's still “ParserError: Source "zeppelin-solidity/contracts/ownership/Ownable.sol" not found: File not supplied initially.”

**Comment by juanfranblanco** [comment]
Do you have at root your node_modules?
![image](https://user-images.githubusercontent.com/562371/37596479-30597040-2b74-11e8-8360-ae2c62648176.png)

also have you set your contract directory as follows?:

![image](https://user-images.githubusercontent.com/562371/37596574-8b109298-2b74-11e8-82bf-d2861249121d.png)

**Comment by qbit86** [comment]
> Do you have at root your node_modules?

Yes:
```
$ ls
build/  contracts/  doc/  migrations/  node_modules/  package-lock.json  test/  truffle.js
```
> also have you set your contract directory as follows?

No, I have no `solidity.*`-related values in user or workspace settings. Adding
```
  "solidity.packageDefaultDependenciesContractsDirectory": ""
```
or
```
  "solidity.packageDefaultDependenciesContractsDirectory": "node_modules"
```
does not help.

**Comment by qbit86** [comment]
@juanfranblanco How to use `"solidity.packageDefaultDependenciesContractsDirectory"` to set path to `node_modules/`?

**Comment by juanfranblanco** [comment]
The contracts directory should be "solidity.packageDefaultDependenciesContractsDirectory": ""
and the 
![image](https://user-images.githubusercontent.com/562371/37700140-41747202-2ce2-11e8-9de7-bfb8ce033b54.png)

**Comment by juanfranblanco** [comment]
Or... set it to contracts "solidity.packageDefaultDependenciesContractsDirectory": "contracts",
![image](https://user-images.githubusercontent.com/562371/37700219-93c17960-2ce2-11e8-8b50-e098fe91f280.png)

**Comment by juanfranblanco** [comment]
the ```node_modules``` is set by default as follows:

![image](https://user-images.githubusercontent.com/562371/37701325-6509df00-2ce6-11e8-8c81-6f9f97e5e1aa.png)

**Comment by juanfranblanco** [comment]
@qbit86  ^^^

**Comment by qbit86** [comment]
Compilation succeeds when I set both `solidity.packageDefaultDependenciesDirectory` and `solidity.packageDefaultDependenciesContractsDirectory` as empty string in workspace settings. But I don't see generated artifacts `*.abi` anymore in root `bin/` folder.

**Comment by juanfranblanco** [comment]
You shouldn't have to set the ``packageDefaultDependenciesDirectory``` to an empty string.  Compilation is as always output to that folder. Have you opened a folder or a file?

**Comment by qbit86** [comment]
@juanfranblanco 

> Have you opened a folder or a file?

I'm opening a folder, but directory for contracts-related stuff is nested, not root. So I have in opened root folder:
```
$ ls
bin/  bps-smart-contracts/  bps-unity-client/  bps-web-app/
```
and in nested contracts-related folder:
```
$ cd ./bps-smart-contracts
$ ls
build/  contracts/  doc/  migrations/  node_modules/  package-lock.json  test/  truffle.js
```

**Comment by qbit86** [comment]
@juanfranblanco 

> You shouldn't have to set the ``packageDefaultDependenciesDirectory``` to an empty string.

Compilation reports success iff both of settings are empty strings (and fails with import error if any of them is not set):
```
{
  "solidity.packageDefaultDependenciesContractsDirectory": "",
  "solidity.packageDefaultDependenciesDirectory": "",
}
```
But when it succeeds I still don't see build artifacts like `*.abi`.

**Comment by W3stside** [comment]
Commenting to verify that @qbit86 's solution has worked for me. Setting just one to a blank string still has path resolving errors. Both however allow for compilation

**Comment by qbit86** [comment]
@W3stside But in my case I only achieved “compilation successful” message, not actual build artifacts (`*.abi` JSON file).

**Comment by W3stside** [comment]
@qbit86 have you tried messing with the `contracts_build_directory` property in `truffle.js`? Haven't tried this but wondering if it may help you

http://truffleframework.com/docs/advanced/configuration

**Comment by qbit86** [comment]
@W3stside Truffle compilation works fine (I'm switching current directory to nested Truffle directory before compiling/migrating). What doesn't work is VS Code extension.

I'm building with Truffle to deploy smart contract; and I'm building with VS Code extension to produce ABI-jsons used on client side.

**Comment by juanfranblanco** [comment]
@qbit86  can you provide a sample structure to test against?

**Comment by qbit86** [comment]
@juanfranblanco 

Branch **develop** in this repository contains minimal repro:
https://github.com/qbit86/solidity-extension-repro

> can you provide a sample structure to test against?

```
.vscode/
    settings.json
repro-smart-contracts/
    build/
    contracts/
        Migrations.sol
        Repro.sol
    migrations/
        1_initial_migration.js
    node_modules/
        ...
    package-lock.json
    truffle.js
```
Files of interest: `.vscode/settings.json` and `repro-smart-contracts/contracts/Repro.sol`.
If set both values in workspace settings to `""` then extension reports successful compilation, but nothing is added to artifacts folder `bin/`.

**Comment by juanfranblanco** [comment]
you need to open the folder repro-smart-contracts instead

**Comment by qbit86** [comment]
@juanfranblanco 

> node_modules is expected at root of the workspace.

Wait, I thought that the whole point of directory setting is that I can hmm... set directory.

> you need to open the folder repro-smart-contracts instead

What does it mean? I'm opening root of my working space; there are several types of sub-projects here (web client, backend, documentation, contracts, etc.), all under the same root, in well defined structure, in one window. C# and C++ extensions do not require that user open only one subdir at time as root. Instead they are working fine if language specific code lives in nested folder, next to folder with Rust or JavaScript code.

**Comment by juanfranblanco** [comment]
I don't disagree, but how do I know your project root, by finding a node_modules directory? Probably if the EIP standard to have a project file is resurrected (the one this issue is actually against) that might be a solution.  Talking about solutions, how do you find your solution? Visual studio has solutions so you can, see what projects form it. This way each project can have its own substructure, including package dependencies that might be in a "javascript" node_modules. 

To me is a priority too, as I want to create projects at solution root anyway. So I will think about it.

**Comment by qbit86** [comment]
@juanfranblanco 

> Visual studio has solutions so you can, see what projects form it

In Visual Studio solutions are auxiliary. In general (at least in case of C#) you can delete `.sln` file, then go down in dir structure to some deeply nested project via command line, execute `msbuild` and it will resolve project's dependencies since they are stored in `.csproj` files, not in `.sln` files.

> but how do I know your project root, by finding a node_modules directory?

I agree, it's one of candidate solutions: traverse up from compiling file location till there is `node_modules/` sibling directory. Or let workspace settings in root `.vscode/settings.json` define set of probing paths (not restricting to `node_modules`) from workspace root. Like this:
```
{
  // "solidity.packageDefaultDependenciesContractsDirectory": "",
  // "solidity.packageDefaultDependenciesDirectory": "",
  "solidity.importProbingPaths": ["repro-smart-contracts/node_modules/"]
}
```

**Comment by juanfranblanco** [comment]
Yes, I thought of that,  but that is not a general setting, it is a project specific. Hence having a project file setting, which will mean looking for a project file within or above the currently selected file. That will be the root. There is already a dappFile, which I think it should be named to Solidity.proj (or something like that).  

The solution file, it becomes probably necessary if you want to compile all projects in the solution.  Mainly  discovery of current projects (not dependencies) and compile each one.

**Comment by NoChance777** [comment]
same issue.

**Comment by juanfranblanco** [comment]
@NoChance777 yes it is not solved yet, but note that this will mean having a project file.
